### PR TITLE
Rename ScriptServiceV3Alpha to KubernetesScriptServiceV1Alpha

### PR DESCRIPTION
**Pull request by APErebus**
# Background

**NOTE:** This replaces #852.

Now that we are no longer relying on the `StartScriptCommandV3Alpha` in the `TentacleClient.ExecuteScript()` contract (#845), we can rename the `ScriptServiceV3Alpha` to `KubernetesScriptServiceV1Alpha` without changing the public contract of `TentacleClient`.

Apart from the rename, this also changes `KubernetesScriptServiceV1Alpha` to only execute scripts via Kubernetes pods.

# Results

The Capabilities service is what tells `TentacleClient` which  services are supported on the target Tentacle. This PR changes this so that if the tentacle is known be running as part of the Kubernetes Agent, then we don't return any of the non-Kubernetes script services (and visa-versa). This gives us a lot more control about what services are advertised for a given tentacle and how TentacleClient uses this.

Unfortunately this PR does remove the tests for the previous `ScriptServiceV3Alpha` because they were executing via local shell, which is not supported in `KubernetesScriptServiceV1Alpha`. We will be adding tests for this new service in the future as it's quite challenging due to requiring a running Kubernetes cluster

## Testing

Tested `TentacleClient` changes against this code agent, older tentacle & this code non-agent tentacle

This PR Agent
image

This PR non-Agent Tentacle Worker
image

Older version non-Agent Tentacle Worker
image

Shortcut story: [sc-73757]

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/main/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [x] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [x] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Octopus.Tentacle.Client/Scripts/ScriptOrchestratorFactory.cs [hunk 1]
```diff
@@ -68,19 +68,22 @@ public async Task<IScriptOrchestrator> CreateOrchestrator(CancellationToken canc
                 throw new OperationCanceledException("Script execution was cancelled", ex);
             }
 
-            return scriptServiceToUse switch
+            if (scriptServiceToUse == ScriptServiceVersion.ScriptServiceVersion1)
```
Comment (APErebus) [inline source/Octopus.Tentacle.Client/Scripts/ScriptOrchestratorFactory.cs:71]: `ScriptServiceToUse` is no longer an enum, so can't be switched on

File: source/Octopus.Tentacle.Client/Scripts/ScriptOrchestratorFactory.cs [hunk 2]
```diff
@@ -127,19 +133,11 @@ async Task<CapabilitiesResponseV2> GetCapabilitiesFunc(CancellationToken ct)
 
             logger.Verbose($"Discovered Tentacle capabilities: {string.Join(",", tentacleCapabilities.SupportedCapabilities)}");
 
-            if (tentacleCapabilities.HasScriptServiceV3Alpha())
+            // Check if we support any kubernetes script service.
+            // It's implied (and tested) that GetCapabilities will only return Kubernetes or non-Kubernetes script services, never a mix
+            if (tentacleCapabilities.HasAnyKubernetesScriptService())
```
Comment (APErebus) [inline source/Octopus.Tentacle.Client/Scripts/ScriptOrchestratorFactory.cs:138]: This is a key feature of this PR, if the Capabilities service returns any Kubernetes script service, we only look for versions in that hierarchy.

If not, we only look for versions of in the `ScriptService` hierarchy

File: source/Octopus.Tentacle.Client/TentacleClientOptions.cs [hunk 3]
```diff
@@ -6,11 +6,6 @@ namespace Octopus.Tentacle.Client
     public class TentacleClientOptions
     {
         public RpcRetrySettings RpcRetrySettings { get; }
-        
-        /// <summary>
-        /// Disables the use of ScriptServiceV3Alpha, even if it's supported on the Tentacle
-        /// </summary>
-        public bool DisableScriptServiceV3Alpha { get; set; }
```
Comment (APErebus) [inline source/Octopus.Tentacle.Client/TentacleClientOptions.cs:13]: We don't need to disable this as the capabilities service now dictates what services are available.

File: source/Octopus.Tentacle.Contracts/KubernetesScriptServiceV1Alpha/StartKubernetesScriptCommandV1Alpha.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        public StartKubernetesScriptCommandV1Alpha(
+            ScriptTicket scriptTicket,
+            string taskId,
+            string scriptBody,
+            string[] arguments,
+            ScriptIsolationLevel isolation,
+            TimeSpan scriptIsolationMutexTimeout,
+            string isolationMutexName,
+            PodImageConfiguration? podImageConfiguration,
```
Comment (APErebus) [inline source/Octopus.Tentacle.Contracts/KubernetesScriptServiceV1Alpha/StartKubernetesScriptCommandV1Alpha.cs:17]: This is a new POCO to hold the pod image configuration (image, url, feed username, password) information, rather than having 4 individual properties

File: source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs [hunk 5]
```diff
@@ -18,8 +20,27 @@ public async Task CapabilitiesAreReturned()
...
-            capabilities.Count.Should().Be(4);
+            capabilities.Count.Should().Be(3);
+
+            capabilities.Should().NotContainMatch("IKubernetesScriptService*");
+        }
+
+        [Test]
+        public async Task OnlyKubernetesScriptServicesAreReturnedWhenRunningAsKubernetesAgent()
+        {
+            Environment.SetEnvironmentVariable(KubernetesConfig.NamespaceVariableName, "ABC");
```
Comment (APErebus) [inline source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs:31]: We need to change the Kubernetes configuration to be testedable without environment variables, but will do that as a separate PR

File: source/Octopus.Tentacle/Services/Capabilities/CapabilitiesServiceV2.cs [hunk 6]
```diff
@@ -15,7 +15,15 @@ public class CapabilitiesServiceV2 : IAsyncCapabilitiesServiceV2
...
-            return new CapabilitiesResponseV2(new List<string>() {nameof(IScriptService), nameof(IFileTransferService), nameof(IScriptServiceV2), nameof(IScriptServiceV3Alpha)});
+
+            //the kubernetes agent only supports the kubernetes script services
+            if (PlatformDetection.Kubernetes.IsRunningAsKubernetesAgent)
+            {
+                return new CapabilitiesResponseV2(new List<string> { nameof(IFileTransferService), nameof(IKubernetesScriptServiceV1Alpha) });
+            }
+
+            //non-kubernetes agent tentacles only support the standard script services
+            return new CapabilitiesResponseV2(new List<string> { nameof(IScriptService), nameof(IFileTransferService), nameof(IScriptServiceV2) });
```
Comment (APErebus) [inline source/Octopus.Tentacle/Services/Capabilities/CapabilitiesServiceV2.cs:26]: The Capabilities response now returns difference script services based on if running as a Kubernetes agent or not

File: source/Octopus.Tentacle/Services/KubernetesServiceAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+
+namespace Octopus.Tentacle.Services
+{
+    [AttributeUsage(AttributeTargets.Class)]
+    public class KubernetesServiceAttribute : Attribute, IServiceAttribute
```
Comment (APErebus) [inline source/Octopus.Tentacle/Services/KubernetesServiceAttribute.cs:6]: The `KubernetesServiceAttribute` which allows us to conditionally register Kubernetes Services... I felt it was better to keep all the services being registered in the same module, rather than splitting them into separate modules

File: source/Octopus.Tentacle/Services/Scripts/Kubernetes/KubernetesScriptServiceV1Alpha.cs [hunk 8]
```diff
@@ -71,33 +71,21 @@ public async Task<ScriptStatusResponseV3Alpha> StartScriptAsync(StartScriptComma
...
                 runningScript.Process = process;
 
-                if (command.DurationToWaitForScriptToFinish != null)
-                {
-                    var waited = Stopwatch.StartNew();
-                    while (process.State != ProcessState.Complete && waited.Elapsed < command.DurationToWaitForScriptToFinish.Value)
-                    {
-                        Thread.Sleep(TimeSpan.FromMilliseconds(10));
-                    }
-                }
```
Comment (APErebus) [inline source/Octopus.Tentacle/Services/Scripts/Kubernetes/KubernetesScriptServiceV1Alpha.cs:88]: This is not on the command and is not supported for the Kubernetes service

**Comment by APErebus** [comment]
Related Server PR: https://github.com/OctopusDeploy/OctopusDeploy/pull/23677

File: .github/CODEOWNERS [hunk 9]
```diff
@@ -3,3 +3,4 @@
 # The Kubernetes parts of Tentacle are owned by Modern Deployments
 /docker/kubernetes-tentacle/ @OctopusDeploy/team-modern-deployments
 /source/Octopus.Tentacle/Kubernetes/ @OctopusDeploy/team-modern-deployments
+/source/Octopus.Tentacle/Services/Scripts/Kubernetes/ @OctopusDeploy/team-modern-deployments
```
Comment (APErebus) [inline .github/CODEOWNERS:6]: I move the renamed service into a sub directory for clarity

File: source/Octopus.Tentacle.Client/Capabilities/CapabilitiesResponseV2ExtensionMethods.cs [hunk 10]
```diff
@@ -18,14 +16,14 @@ public static bool HasScriptServiceV2(this CapabilitiesResponseV2 capabilities)
             return capabilities.SupportedCapabilities.Contains(nameof(IScriptServiceV2));
         }
 
-        public static bool HasScriptServiceV3Alpha(this CapabilitiesResponseV2 capabilities)
+        public static bool HasAnyKubernetesScriptService(this CapabilitiesResponseV2 capabilities)
```
Comment (LukeButters) [inline source/Octopus.Tentacle.Client/Capabilities/CapabilitiesResponseV2ExtensionMethods.cs:19]: Is it worth pulling this out into some sort of k8s directory owned by your team? So you don't need SaST approval for future changes to this?

File: source/Octopus.Tentacle.Client/Scripts/ScriptOrchestratorFactory.cs (hunk 2 above)
Comment (LukeButters) [reply source/Octopus.Tentacle.Client/Scripts/ScriptOrchestratorFactory.cs:138]: Also +1 to this method with a single if calling to well named methods.

File: source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs (hunk 5 above)
Comment (LukeButters) [reply source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs:31]: Ideally yes but many of the `Tentacle.Tests` (distinct from integration tests) suffer from environment variables being used e.g. within tentacle when a script is run :( 

Its frankly a bit of a mess.

**Review by LukeButters — APPROVED** [review]

File: source/Octopus.Tentacle.Client/Capabilities/CapabilitiesResponseV2ExtensionMethods.cs (hunk 10 above)
Comment (APErebus) [reply source/Octopus.Tentacle.Client/Capabilities/CapabilitiesResponseV2ExtensionMethods.cs:19]: Yep, I can do that :)

Comment (APErebus) [reply source/Octopus.Tentacle.Client/Capabilities/CapabilitiesResponseV2ExtensionMethods.cs]: Moved in 0dff6b00

File: source/Octopus.Tentacle.Tests.Integration/CapabilitiesServiceV2Test.cs [hunk 11]
```diff
@@ -32,17 +32,38 @@ public async Task CapabilitiesFromAnOlderTentacleWhichHasNoCapabilitiesService_W
...
+
+            //all versions have ScriptServiceV1 & IFileTransferService
+            var expectedCapabilitiesCount = 2;
             if (version.HasScriptServiceV2())
             {
                 capabilities.Should().Contain("IScriptServiceV2");
                 expectedCapabilitiesCount++;
             }
 
+            capabilities.Should().NotContain("IKubernetesServiceV1Alpha");
```
Comment (flin-8) [inline source/Octopus.Tentacle.Tests.Integration/CapabilitiesServiceV2Test.cs:65]: Should we make this more general? match on anything with `Kubernetes` in it?

File: source/Octopus.Tentacle/Services/ServicesModule.cs [hunk 12]
```diff
@@ -22,19 +21,21 @@ protected override void Load(ContainerBuilder builder)
 
             builder.RegisterType<NuGetPackageInstaller>().As<IPackageInstaller>();
 
-            // Register the script executor based on if we should
+            // Register our Halibut services
+            var allTypes = ThisAssembly.GetTypes();
+            RegisterHalibutServices<ServiceAttribute>(builder, allTypes);
+
+            //only register kubernetes services when
```
Comment (flin-8) [inline source/Octopus.Tentacle/Services/ServicesModule.cs:28]: We are still registering non-K8s services when running as a an Agent.

Should we consider registering the services explicitly that's somehow in sync with what `CapabilitiesServiceV2` returns

File: source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs [hunk 13]
```diff
@@ -18,8 +20,27 @@ public async Task CapabilitiesAreReturned()
             capabilities.Should().Contain("IScriptService");
             capabilities.Should().Contain("IFileTransferService");
             capabilities.Should().Contain("IScriptServiceV2");
-            capabilities.Should().Contain("IScriptServiceV3Alpha");
-            capabilities.Count.Should().Be(4);
+            capabilities.Count.Should().Be(3);
```
Comment (flin-8) [inline source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs:23]: can all these assertions be rewritten as 

`capabilities.Should().BeEquivalentTo("IScriptService", "IFileTransferService", "IScriptServiceV2");`

**Review by flin-8 — APPROVED** [review]
LGTM + comments

File: source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs (hunk 13 above)
Comment (APErebus) [reply source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs:23]: Yes they can. They can also use `nameof()` so I'll change to use that

File: source/Octopus.Tentacle.Tests.Integration/CapabilitiesServiceV2Test.cs (hunk 11 above)
Comment (APErebus) [reply source/Octopus.Tentacle.Tests.Integration/CapabilitiesServiceV2Test.cs]: I think having it specific makes sense for now. If we get another Kubernetes service, we can re-evaluate it

File: source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs (hunk 13 above)
Comment (APErebus) [reply source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs:24]: fixed in [8f59057](https://github.com/OctopusDeploy/OctopusTentacle/pull/859/commits/8f59057fe87be8cb8c032d04ec46aa801b3ef9b9)

File: source/Octopus.Tentacle/Services/ServicesModule.cs (hunk 12 above)
Comment (APErebus) [reply source/Octopus.Tentacle/Services/ServicesModule.cs:28]: Might be worth discussing with @OctopusDeploy/team-server-at-scale at some point 🤔
